syntax = "proto3";
package coop.rchain.comm.protocol.routing;

import "scalapb/scalapb.proto";
import "google/protobuf/any.proto";

option (scalapb.options) = {
  package_name: "coop.rchain.comm.protocol.routing"
  flat_package: true
};

message Node {
    bytes  id       = 1;
    bytes  host     = 2;
    uint32 tcp_port = 3;
    uint32 udp_port = 4;
}

message Header {
  Node   sender         = 1;
}

message Heartbeat {
}

message HeartbeatResponse {
}

message ProtocolHandshake {
  bytes nonce       = 1;
}

message ProtocolHandshakeResponse {
  bytes nonce       = 1;
}

message Packet {
  string typeId  = 1;
  bytes  content = 2;
}

message Disconnect {
}

message Protocol {
    Header header                                                 = 1;
    oneof message {
        Heartbeat                   heartbeat                     = 2;
        HeartbeatResponse           heartbeat_response            = 3;
        ProtocolHandshake           protocol_handshake            = 4;
        ProtocolHandshakeResponse   protocol_handshake_response   = 5;
        Packet                      packet                        = 6;
        Disconnect                  disconnect                    = 7;
    }
}

service TransportLayer {
  rpc Tell (TLRequest) returns (TLResponse) {}
  rpc Ask (TLRequest) returns (TLResponse) {}
}

message TLRequest {
  Protocol protocol = 1;
}

message InternalServerError {
  bytes error = 1;
}

message NoResponse {
}

message TLResponse {
  oneof payload {
    Protocol protocol                       = 1;
    NoResponse noResponse                   = 2;
    InternalServerError internalServerError = 3;
  }
}
